simplify compile method's code

aidewoode opened the pull request:
Hi,  I noticed the original code ```char.tr('A-Z', 'a-z') ``` was to downcase the capital letter ， So why not just use ```String#downcase``` method?

kytrinyx commented:
Interesting--any idea if there's a significant difference in performance between the two? I'd be curious to know if the original author said something specific about choosing `tr`.

davydovanton commented:
@kytrinyx I [created benchmark](https://gist.github.com/davydovanton/c47f9a33f19fd34582c3) and `downcase` is faster than `tr` :+1:

davydovanton commented on lib/sinatra/base.rb:
```diff
@@ -1691,7 +1691,7 @@ def safe_ignore(ignore)
         unsafe_patterns = unsafe_ignore.map! do |unsafe|
           chars = unsafe.split(//).map! do |char|
             if char =~ /[A-Z]/
```
also I think that this and previous `if` unnecessary (if we use `downcase` here). @aidewoode, @kytrinyx what do you think about this?

aidewoode commented:
@davydovanton you are right, if we can use downcase,  the ```if``` is unnecessary. Because the ```downcase``` method only affect on A-Z letters. Now I just wondering why the original author use the ```tr``` method? He maybe make a mistake or have specific purpose.

kgrz commented:
I think the`if` is necessary here even if downcase is used.

With the `if` clause, "Hello" would result in `["Hh", "e", "l", "l", "o"]`

It appends a second character if it's a capital letter. Without the `if`, that logic would result in all characters in array having two items.

aidewoode commented:
@kgrz I made a mistake, you are right, the ```if ``` is necessary, but we can write code don't use ```if ```,
like this

```ruby
a = "A"
b = a.dup
b <<= a.downcase! || ""  #=> "Aa"
```

kytrinyx commented:
`a.downcase!` would never return `nil`, right?

aidewoode commented:
@kytrinyx  yeah,  ```a.downcase!``` would never return nil, but when a is not a A-Z letter, it will return nil. and I'm sorry , I made a mistake the ``` b << a.downcase! || ''``` should write like this ```b << (a.downcase! || '')```, So the original code 

```ruby
chars = unsafe.split(//).map! do |char|
  if char =~ /[A-Z]/
     char <<= char.tr('A-Z', 'a-z')
  end
  char
end
```

could write like this:
```ruby
chars = unsafe.split(//).map! do |char|
  char_dup = char.dup
  char_dup << (char.downcase! || '')
end
````

kgrz commented:
Not sure what I'm missing, but isn't

```ruby
  char_dup = char.dup
  char_dup << (char.downcase! || '')
```
what `char.downcase` does (dup the string, and downcase it). Is there a perf advantage of removing the `if` condition?

southwolf commented:
@aidewoode  The `if` way is easier to understand, maybe?

aidewoode commented:
@kgrz @southwolf   yeah, I know , the ```if ``` is more straightforward, but the ```downcase``` method only affect on A-Z character. So, I think if we can use ```downcase``` there, why we should determine the the character belongs to A-Z. And I think this just a insignificant thing, So we don't have to discuss the ```if ``` statement anymore.

msg7086 commented:
Maybe `{ |char| char + (char.dup.downcase! || '') }` ?

aidewoode commented:
@msg7086 Thank you, I will change the code.

kytrinyx commented:
Thanks, everyone for taking the time to discuss this so thoroughly.
